src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalStableLayoutManager.java

```diff
@@ -0,0 +1,1913 @@
+                    e.relativeFrom = l.getFrom().getRelativePosition().x;
+                    e.relativeTo = l.getTo().getRelativePosition().x;
+                    node.preds.add(e);
+                } else if (l.getFrom().getVertex().equals(action.vertex)
+                        && nodes.contains(vertexToLayoutNode.get(l.getTo().getVertex()))) {
+                    e.from = node;
+                    e.to = vertexToLayoutNode.get(l.getTo().getVertex());
+                    e.relativeFrom = l.getFrom().getRelativePosition().x;
+                    e.relativeTo = l.getTo().getRelativePosition().x;
+                    node.succs.add(e);
+                }
+            }
+            insertNode(node, layer);
+            node.succs.clear();
+            node.preds.clear();
+
+            // Add associated edges
+            for (LinkAction a : action.linkActions) {
+                if (a.action == Action.ADD) {
+                    applyAddLinkAction(a.link);
+                }
+            }
+
+            sanityCheckEdges();
+            sanityCheckNodesAndLayerNodes();
+        }
+
+        private void applyRemoveLinkAction(Link l) {
+            Vertex from = l.getFrom().getVertex();
+            Vertex to = l.getTo().getVertex();
+            LayoutNode toNode = vertexToLayoutNode.get(to);
+            LayoutNode fromNode = vertexToLayoutNode.get(from);
+
+            if (toNode.layer < fromNode.layer) {
+                // Reversed edge
+                LayoutNode temp = toNode;
+                toNode = fromNode;
+                fromNode = temp;
+
+                reversedLinks.remove(l);
+                reversedLinkEndPoints.remove(l);
+                reversedLinkStartPoints.remove(l);
+            }
+
+            // Remove preds-edges bottom up, starting at "to" node
+            // Cannot start from "from" node since there might be joint edges
+            List<LayoutEdge> toNodePredsEdges = List.copyOf(toNode.preds);
+            for (LayoutEdge edge : toNodePredsEdges) {
+                LayoutNode n = edge.from;
+                LayoutEdge edgeToRemove;
+
+                if (edge.link != null && edge.link.equals(l)) {
+                    toNode.preds.remove(edge);
+                    edgeToRemove = edge;
+                } else {
+                    // Wrong edge, look at next
+                    continue;
+                }
+
+                if (n.vertex != null && n.vertex.equals(from)) {
+                    // No dummy nodes inbetween 'from' and 'to' vertex
+                    n.succs.remove(edgeToRemove);
+                    break;
+                } else {
+                    // Must remove edges between dummy nodes
+                    boolean found = true;
+                    LayoutNode prev = toNode;
+                    while (n.vertex == null && found) {
+                        found = false;
+
+                        if (n.vertex == null && n.succs.size() <= 1 && n.preds.size() <= 1) {
```

Comment:
This is to make the loop break once we hit the non-dummy node where the edge goes from. I.e. for the edge (u,v) with lots of dummy nodes in between nodes u and v, we only want to remove the dummy nodes and then break the loop as soon as we are at node u.

Reply:
Exactly, if `n.vertex == null` we want to break the loop. There are two cases we need to consider when removing the dummy nodes: 1) there is a long chain of dummy nodes between node u and v, and 2) the edge is part of an edge concentration of multiple edges. In case 1) we just traverse the edge with all the dummy nodes and remove them as we go until we hit node u (`n.vertex != null`). In case 2) we traverse along the edge until we find the anchor node (which is a dummy node with one ingoing edge and multiple outgoing edges), and break the loop when we reach the anchor node.

Reply:
Thanks for explaining it in more details. We probably had a misunderstanding here, though. Toby and I were referring to this `n.vertex == null` here:
```
 if (n.vertex == null && n.succs.size() <= 1 && n.preds.size() <= 1)
```
and not the one in
```
while (n.vertex == null && found) 
```
From your explanation it makes sense to keep the one in the `while` but the other one in the `if` always seems to be true and could thus be removed.

Reply:
Oh okay thanks for clarifying that makes sense!